rfcs/NNNN-preact-signals.md

```diff
@@ -0,0 +1,184 @@
+      // also bail early so we don't erroneously appear to not depend on any
+      // signals.
+      if (!this.isUpdatePending) {
+        return;
+      }
+      // If we have a previous effect, dispose it
+      this._disposeEffect?.();
+
+      // We create a new effect to capture all signal access within the
+      // performUpdate phase (update, render, updated, etc) of the element.
+      this._disposeEffect = effect(() => {
+        // When Signals change we need to re-render, but we need to get past
+        // the isUpdatePending in performUpdate(), so we set it to true.
+        this.isUpdatePending = true;
+        // We call super.performUpdate() so that we don't create a new effect
+        // only as the result of the effect running.
+        super.performUpdate();
+      });
+    }
+```
+
+An issue with this approach is that we bypass the reactive update lifecycle when signals change. We would like to integrate updates from signals and reactive properties into one batch. To do this we need separate watch and update code paths. Preact Signals don't have such an API, but we should be able to simulate it:
+
+```ts
+    override performUpdate() {
+      if (!this.isUpdatePending) {
+        return;
+      }
+      this._disposeEffect?.();
+      let updateFromLit = true;
+      this._disposeEffect = effect(() => {
+        if (updateFromLit) {
+          super.performUpdate();
+        } else {
+          // This branch is an effect run from Preact signals.
+          // This will cause another call into performUpdate, which will
+          // then create a new effect watching that update pass.
+          this.requestUpdate();
+        }
+        updateFromLit = false;
+      });
+    }
+```
+
+### watch() directive
+
+The `watch()` async directive accepts a signal and renders its value synchronously to the containing binding. When the signal changes, the binding value is updated directly.
+
+Usage:
+```ts
+html`<p>${watch(messageSignal)}</p>`
+```
+
+```ts
+class WatchSignal extends AsyncDirective {
+  render(signal: Signal<unknown>) {
+    let updateFromLit = true;
+    effect(() => {
+      // Access the signal unconditionally to watch it
+      const value = signal.value;
+      if (!updateFromLit) {
+        this.setValue(value);
+      }
+      updateFromLit = false;
+    });
+    return signal.value;
+  }
+}
+export const watch = directive(WatchSignal);
+```
+
+### Auto-watching template tag
+
+Auto-watching versions of `html` and `svg` template tags will scan a template result's values and automatically wrap them in a `watch()` directive if they are signals.
+
+We should be able to detect signals with `value instanceof Signal`. Which the `instanceof` operator is fragile, especially in the presence of multiple copies of a module, it appears that the Preact Signals modules are _already_ fragile in this respect due to module-scoped state: all signals, computed signals, and effects, must use the same module instance to work.
```

Comment:
Per discussion, add a note that we will request a brand on preact signals for less-brittle checks?

Reply:
done